crates/cargo-util/src/process_builder.rs

```diff
@@ -411,21 +412,17 @@ impl ProcessBuilder {
         log::debug!("created argfile at {path} for `{self}`");
 
         let cap = self.get_args().map(|arg| arg.len() + 1).sum::<usize>();
-        let mut buf = String::with_capacity(cap);
+        let mut buf = Vec::with_capacity(cap);
         for arg in &self.args {
-            let arg = arg
-                .to_str()
-                .ok_or_else(|| {
-                    io::Error::new(
-                        io::ErrorKind::Other,
-                        "argument contains invalid UTF-8 characters",
-                    )
-                })?;
-            // TODO: Shall we escape line feed?
-            buf.push_str(arg);
-            buf.push('\n');
+            let arg = arg.to_str().ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::Other,
+                    "argument contains invalid UTF-8 characters",
+                )
+            })?;
+            writeln!(buf, "{}", ArgEscape::from(arg))?;
```

Comment:
I guess the other question to escaping is what do rustc and similar tools support?

Reply:
This is [how rustc generates argfiles for ld](https://github.com/rust-lang/rust/blob/8c1fb2eb23964b56bd279b12aa925935d8559870/compiler/rustc_codegen_ssa/src/back/link.rs#L1442-L1480). As I read it so far, they just follow the specification. rustc uses only newlines to determine so I guess escaping `\r` `\n` and `\\` might be sufficient.

However, I found that [rustc_driver just naively reads line by line from argfile](https://github.com/rust-lang/rust/blob/1.60.0/compiler/rustc_driver/src/args.rs#L7-L18), they won't do any unescaping. As long as rustc does nothing special and our primary use case is serving rustc, I decide to **remove the escape part and error out** if someone passes line feeds in, as you recommended earlier. 

BTW there is [a test case](https://github.com/rust-lang/cargo/blob/5f400f098ba6bd76eb57c11bbe7c39a443146afe/tests/testsuite/rustdocflags.rs#L101-L123) validating that behaviour, which unfortunately accept whitespaces in `RUSTDOCFLAGS`.